crates/aptos-crypto/src/secp256k1.rs

```diff
@@ -0,0 +1,300 @@
+    }
+}
+
+impl traits::Uniform for PrivateKey {
+    fn generate<R>(rng: &mut R) -> Self
+    where
+        R: ::rand::RngCore + ::rand::CryptoRng + ::rand_core::CryptoRng + ::rand_core::RngCore,
+    {
+        loop {
+            let mut ret = [0u8; PRIVATE_KEY_LENGTH];
+            rng.fill_bytes(&mut ret);
+            if let Ok(key) = libsecp256k1::SecretKey::parse(&ret) {
+                return Self(key);
+            }
+        }
+    }
+}
+
+impl ValidCryptoMaterial for PrivateKey {
+    fn to_bytes(&self) -> Vec<u8> {
+        self.to_bytes().to_vec()
+    }
+}
+
+/// SecP256K1 public key
+#[derive(DeserializeKey, Clone, Eq, PartialEq, SerializeKey)]
+pub struct PublicKey(pub(crate) libsecp256k1::PublicKey);
+
+impl PublicKey {
+    /// Serialize the public key into a byte vector (full length)
+    pub fn to_bytes(&self) -> Vec<u8> {
+        self.0.serialize().to_vec()
+    }
+}
+
+impl std::fmt::Debug for PublicKey {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "secp256k1::PublicKey({})", self)
+    }
+}
+
+impl std::fmt::Display for PublicKey {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", hex::encode(&self.to_bytes()[..]))
+    }
+}
+
+impl std::hash::Hash for PublicKey {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        let encoded_public_key = self.to_bytes();
+        state.write(&encoded_public_key);
+    }
+}
+
+impl TryFrom<&[u8]> for PublicKey {
+    type Error = CryptoMaterialError;
+
+    fn try_from(bytes: &[u8]) -> std::result::Result<PublicKey, CryptoMaterialError> {
+        match libsecp256k1::PublicKey::parse_slice(bytes, None) {
+            Ok(public_key) => Ok(PublicKey(public_key)),
+            Err(_) => Err(CryptoMaterialError::DeserializationError),
+        }
+    }
+}
+
+impl From<&PrivateKey> for PublicKey {
+    fn from(private_key: &PrivateKey) -> Self {
+        PublicKey(libsecp256k1::PublicKey::from_secret_key(&private_key.0))
+    }
+}
+
+impl traits::PublicKey for PublicKey {
+    type PrivateKeyMaterial = PrivateKey;
+}
+
+impl traits::Length for PublicKey {
+    fn length(&self) -> usize {
+        PUBLIC_KEY_LENGTH
+    }
+}
+
+impl ValidCryptoMaterial for PublicKey {
+    fn to_bytes(&self) -> Vec<u8> {
+        self.to_bytes().to_vec()
+    }
+}
+
+impl traits::VerifyingKey for PublicKey {
+    type SignatureMaterial = Signature;
+    type SigningKeyMaterial = PrivateKey;
+}
+
+/// SecP256K1 signature
+#[derive(DeserializeKey, Clone, SerializeKey)]
+pub struct Signature(pub(crate) libsecp256k1::Signature);
+
+impl Signature {
+    /// Serialize the signature into a byte vector
+    pub fn to_bytes(&self) -> Vec<u8> {
+        self.0.serialize().to_vec()
+    }
+
+    fn verify(
+        &self,
+        message: &libsecp256k1::Message,
+        public_key: &libsecp256k1::PublicKey,
+    ) -> Result<()> {
+        if libsecp256k1::verify(message, &self.0, public_key) {
```

Comment:
High `s` check **MUST** be done here.

Reply:
There is no way for a high s to get here if we block it at the deserialization time.